src/interrogate/coverage.py

```diff
@@ -459,21 +464,36 @@ def _sort_results(results: InterrogateResults) -> InterrogateResults:
         all_filenames_map = {r.filename: r for r in results.file_results}
         all_dirs = sorted(
             {
-                os.path.dirname(r.filename)
+                r.filename.parent
                 for r in results.file_results
-                if os.path.dirname(r.filename) != ""
+                if r.filename.parent
             }
         )
 
-        sorted_results = []
+        # cnt = 0
+
+        # def log(obj: object) -> None:
+        #     nonlocal cnt
+        #     cnt += 1
+        #     pth = Path.home() / "dev/download/interrogate/log.txt"
+        #     with pth.open("a") as f:
+        #         f.write(f"{cnt} {obj!r}\n")
+
+        # log(all_filenames_map)  # 1
+        # log(all_filenames_map_paths)  # 2
+        # log(all_dirs)  # 3
+        # log(all_dirs_paths)  # 4
+
+        sorted_results: list[Path] = []
         while all_dirs:
             current_dir = all_dirs.pop(0)
-            files = []
-            for p in os.listdir(current_dir):
-                path = os.path.join(current_dir, p)
-                if path in all_filenames_map.keys():
-                    files.append(path)
-            files = sorted(files)
+            files = sorted(
+                [
+                    path
+                    for path in current_dir.iterdir()
+                    if path in all_filenames_map
+                ]
+            )
```

Comment:
```suggestion
            files = [
                path
                for path in current_dir.iterdir()
                if path in all_filenames_map
            ]
            files.sort()
```
Do not create an intermediate `list`, which is sorted by `sorted`, because this is sorting on iterables, but not inplace sorting. Afterwards, it creates another list.

sorted: https://docs.python.org/3/library/functions.html#sorted  
list-sort: https://docs.python.org/3/library/stdtypes.html#list.sort

Also the original code was not good, because it doesn't use inplace sorting.